### DISCUSSION TITLE
Should the generator use more specific SafeHandles where possible?

### DISCUSSION THREAD

**Discussion by HavenDV** [microsoft/CsWin32#691] [General]
For example: PInvoke.CreateNamedPipe can return SafePipeHandle instead SafeFileHandle. This is in the same namespace and library

**Comment by AArnott** [comment]
This is a tough question. On Windows, `SafeFileHandle` and `SafePipeHandle` are equivalent, in that they both call kernel32's `CloseHandle`. All we know from the win32metadata is that certain APIs produce handles that are released by CloseHandle. If we did the same thing we do in CsWin32 in general for handles, we would create a `CloseHandleSafeHandle` class of our own and use that. But we try to reuse .NET APIs where we can for better interop, so we have a special line to reuse `SafeFileHandle` anywhere that `CloseHandle` is prescribed. 
This means that CsWin32 will always use that SafeHandle type for CloseHandle, even for Pipe-related APIs, and where interop with the BCL would prefer a different SafeHandle type like SafePipeHandle. 

There's no great way to handle this. But suffice it to say, this isn't a _bug_, but it's possible to be inconvenient at times when you want to create a pipe and hand it off to another .NET API that needs a SafeHandle by a different type.
You can always convert the SafeFileHandle though. Maybe something like this (untested):

```cs
SafePipeHandle pipeHandle = new(fileHandle.DangerousGetHandle());
fileHandle.Dispose();
```

The call to `DangerousGetHandle` should increment a counter such that disposing of the file handle will *not* actually call `CloseHandle`, thus allowing the `pipeHandle` to now control the lifetime of the handle.

**Reply by HavenDV** [reply]
Got it, thanks, it was just a suggestion. I just used the overload that returns `HANDLE` and created a `SafePipeHandle` from that to pass into the `NamedPipeServerStream` constructor
